packages/labs/analyzer/src/lib/javascript/mixins.ts

```diff
@@ -0,0 +1,205 @@
+  // a class expression, hence why we effectively disallow them here. However,
+  // we may want to loosen this constraint for JS users some day
+  if (!analyzer.typescript.isBlock(functionBody)) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to have a block function body; arrow-function class ` +
+        `expression syntax is not supported.`,
+      analyzer
+    );
+    return undefined;
+  }
+  let classDeclaration: ts.ClassDeclaration | null = null;
+  let returnStatement: ts.ReturnStatement | null = null;
+  for (const s of functionBody.statements) {
+    if (analyzer.typescript.isClassDeclaration(s)) {
+      classDeclaration = s;
+    }
+    if (analyzer.typescript.isReturnStatement(s)) {
+      returnStatement = s;
+    }
+  }
+  if (!classDeclaration) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain a class declaration statement.`,
+      analyzer
+    );
+    return undefined;
+  }
+  if (!returnStatement) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain a return statement returning a class.`,
+      analyzer
+    );
+    return undefined;
+  }
+  const extendsClause = classDeclaration.heritageClauses?.find(
+    (c) => c.token === analyzer.typescript.SyntaxKind.ExtendsKeyword
+  );
+  if (extendsClause === undefined) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain class declaration extending a superClass argument to function.`,
+      analyzer
+    );
+    return undefined;
+  }
+  if (extendsClause.types.length !== 1) {
+    analyzer.addDiagnostic(
+      createDiagnostic({
+        typescript: analyzer.typescript,
+        node: extendsClause,
+        message:
+          'Internal error: did not expect a mixin class extends clause to have multiple types',
+        code: DiagnosticCode.UNSUPPORTED,
+        category: analyzer.typescript.DiagnosticCategory.Warning,
+      })
+    );
+    return undefined;
+  }
+  const superClassArgIdx = findSuperClassArgIndexFromHeritage(
+    possibleSuperClasses,
+    extendsClause.types[0].expression,
+    analyzer
+  );
+  if (superClassArgIdx < 0) {
+    analyzer.addDiagnostic(
+      createDiagnostic({
+        typescript: analyzer.typescript,
+        node: extendsClause,
+        message:
+          'Did not find a "superClass" argument used in the extends clause of mixin class.',
+        code: DiagnosticCode.UNSUPPORTED,
+        category: analyzer.typescript.DiagnosticCategory.Warning,
+      })
+    );
+    return undefined;
+  }
+
+  const classDeclarationName = classDeclaration.name?.text ?? name;
+
+  return {
+    node: fn,
+    name,
+    superClassArgIdx,
+    classDeclaration: getClassDeclaration(
+      classDeclaration,
+      classDeclarationName,
+      analyzer,
+      undefined,
+      true /* isMixinClass */
+    ),
+  };
+};
+
+const findSuperClassArgIndexFromHeritage = (
```

Comment:
Is this finding the index in the mixin function's parameter list or the heritage call expression's argument list?

Note that in general, the superclass arg could be a named property too. I'm not sure what TypeScript's rules are.

Reply:
it is the index in the function's parameter list, based on the heritage of the class

i think if we want to change this, we should probably do it in a follow-up since it could be a bit of a refactor to allow an object to be used as a parameter

Reply:
I'm not sure we _should_ fix this, especially if TypeScript is strict about recognizing mixins. We can keep an eye on it.